source/slang/slang-ir-spirv-legalize.cpp

```diff
@@ -254,8 +254,30 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
             }
             else
             {
-                if (!as<IRDecoration>(use->getUser()))
+                auto user = use->getUser();
+                if (!as<IRDecoration>(user))
                 {
+                    // do not emit load if we have a `store(Ptr<T>, globalParam<T>)`, instead
```

Comment:
This should be done in a separate pass. I believe we already have a `removeRedundantLoadStore` pass that tries to fold loads and stores, and it has more condition checks to ensure the operation can be done safely.

We shouldn't need to implement this logic here. If this is to workaround some problems, we need to tackle that issue in a different way.

Reply:
this is to solve the previous issue with textures. I will check if the logic can be moved to `removeRedundantLoadStore`

I noticed if we use a constref/ref we still have an issue with `insertLoadAtLatestLocation` inserting a Load it should not insert. This also was why interpolation was breaking.

For textures specifically we have the additional issue that we cannot just store(dst,load(src)) since we have a special way of constructing the type.

Reply:
I think the general logic here is to not insert a load when the pointer is being used in a call, and that call takes an `inout` or `ref` parameter at the argument index.

Reply:
this would work for most cases, interpolate does exactly this (its exactly the solution I used)

for TextureType since it is not an L value I must use a `__constref` which does not appear as a ref in SPIR-V legalization

adding a new step to eliminateRedundantLoadStore does fix the issue from what it looks like (and adding the step to run after SPIR-V simplification)